### DISCUSSION TITLE
azure pipeline agents don't always remove

### DISCUSSION THREAD

**Discussion by BenH-Puregym** [kedacore/keda#3349] [Q&A / Need Help]
### Report

i've got scaledjob setup and a placeholder agent in ADO. Sometimes the agents are remove when their job is finished and sometimes they're not. Some agents which don't get removed will have a job lasting 13 minutes and others 1hr so doesn't seem to be consistent.  

### Expected Behavior

the agents should remove each time a job has finished

### Actual Behavior

some do, some don't. leaving the pool with tons of offline agents.

### Steps to Reproduce the Problem

1. setup a pipeline scaledjob with 10/20 containers
2. run a unit test pipeline or similar which has 10 parallel jobs
3. repeat a few times and you will see the agents being left.

### Logs from KEDA operator

```
example
```

### Platform

Microsoft Azure

### Scaler Details

pipeline

**Comment by JorTurFer** [comment]
hey,
Is the agent pod finishing? I mean, KEDA doesn't know if the pod has finished or not its work, KEDA removes the pod/job once it has finished with success or failure, so if your pod is still alive without finishing (like a job should do), KEDA will not remove it

**Comment by BenH-Puregym** [comment]
Hey  @JorTurFer , So the agent is left in an offline state in the ADO agent pool but when you check back in AKS, the pod and job is no longer there or the job shows as success. Then requires to be manually removed from ADO.

**Comment by JorTurFer** [comment]
Just to clarify it, is it your problem related with agent pods not deleted from the AKS or with orphan agent in Azure DevOps?

**Comment by BenH-Puregym** [comment]
maybe would be best for me to explain the problem again:
1. job is run on ADO
2. keda starts jobs and adds pods to the ADO pool
3. job finishes on ADO
4. keda deletes the pod and stops the job
5. agents still exist in ADO pool as offline

**Reply by tomkerkhove** [reply]
So what @JorTurFer says in [#3349 (comment)](https://github.com/kedacore/keda/discussions/3349?converting=1#discussioncomment-3118459) is correct - KEDA just adds instances to an existing ADO pool to spin up pipelines based on the current pipeline queue.

However, to be sure - Would you mind adding some screenshots from Azure DevOps/Pipelines to see if it works as intended please? Ideally just of your ADO pool and its overview

**Reply by BenH-Puregym** [reply]
so perhaps a problem with the script that MS have provided [here ](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops) which is supposed to remove the agent when the job is finished rather than a problem with Keda. Like I say, it work sometimes but doesn't always.

screenshot as requests:
![image](https://user-images.githubusercontent.com/89586825/178221251-d7bb0bcd-570c-4dc2-9f94-b0d43a2ff4ba.png)

as you can see, some of them didn't even run a job before going offline.

**Reply by tomkerkhove** [reply]
That seems like a agent problem rather than KEDA indeed - Thanks for sharing!

Once a job completes, it should remove it from the pool indeed.

**Comment by JorTurFer** [comment]
hey, 
I got it. KEDA doesn't handle ADO at all, the only goal of KEDA is to make the autoscaling dead simple, but not to manage the upstreams. 
I mean, in the same way that you register the agent by yourself in AzDO (it does automatically on pod agent) you should unregister it. My suggestion is to unregister the agent after executing the pipeline, before finishing the job execution.

**Comment by kamilpi** [comment]
Hi,

I think you do not need to unregister it from ADO.

I leave it here for future similar issues.
It depends on how you build the image of the ADO agent.
According to the documentation you can find the flag that will shut down the agent after the build is executed: https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops#start-the-image

Script `start.sh`:
```
./run.sh --once "$@" & wait $!
```

Enjoy!
